frame/evm/src/resource.rs

```diff
@@ -0,0 +1,627 @@
+
+#[derive(Debug, PartialEq)]
+/// Resource error.
+pub enum ResourceError {
+	/// The Resource usage exceeds the limit.
+	LimitExceeded,
+	/// Invalid Base Cost.
+	InvalidBaseCost,
+	///Used to indicate that the code should be unreachable.
+	Unreachable,
+}
+
+impl From<ResourceError> for ExitError {
+	fn from(_error: ResourceError) -> Self {
+		ExitError::OutOfGas
+	}
+}
+
+/// A struct that keeps track of resource usage and limit.
+pub struct Resource<T> {
+	limit: T,
+	usage: T,
+}
+
+impl<T> Resource<T>
+where
+	T: CheckedAdd + Saturating + PartialOrd + Copy,
+{
+	/// Creates a new `Resource` instance with the given base cost and limit.
+	///
+	/// # Errors
+	///
+	/// Returns `ResourceError::InvalidBaseCost` if the base cost is greater than the limit.
+	pub fn new(base_cost: T, limit: T) -> Result<Self, ResourceError> {
+		if base_cost > limit {
+			return Err(ResourceError::InvalidBaseCost);
+		}
+		Ok(Self {
+			limit,
+			usage: base_cost,
+		})
+	}
+
+	/// Records the cost of an operation and updates the usage.
+	///
+	/// # Errors
+	///
+	/// Returns `ResourceError::LimitExceeded` if the Resource usage exceeds the limit.
+	fn record_cost(&mut self, cost: T) -> Result<(), ResourceError> {
+		let usage = self
+			.usage
+			.checked_add(&cost)
+			.ok_or(ResourceError::LimitExceeded)?;
+
+		if usage > self.limit {
+			return Err(ResourceError::LimitExceeded);
+		}
+		self.usage = usage;
+		Ok(())
+	}
+
+	/// Refunds the given amount.
+	fn refund(&mut self, amount: T) {
+		self.usage = self.usage.saturating_sub(amount);
+	}
+
+	/// Returns the usage.
+	fn usage(&self) -> T {
+		self.usage
+	}
+}
+
+pub enum AccessedStorage {
+	AccountCodes(H160),
+	AccountStorages((H160, H256)),
+}
+
+#[derive(Default, Clone, Eq, PartialEq)]
+pub struct Recorded {
+	account_codes: Vec<H160>,
+	account_storages: BTreeMap<(H160, H256), bool>,
+}
+
+/// A struct that keeps track of the proof size and limit.
+pub struct ProofSizeResource<T> {
+	resource: Resource<u64>,
+	recorded: Recorded,
+	_marker: PhantomData<T>,
+}
+
+impl<T: Config> ProofSizeResource<T> {
+	/// Creates a new `ProofSizeResource` instance with the given limit.
+	pub fn new(base_cost: u64, limit: u64) -> Result<Self, ResourceError> {
+		Ok(Self {
+			resource: Resource::new(base_cost, limit)?,
+			recorded: Recorded::default(),
+			_marker: PhantomData,
+		})
+	}
+
+	/// Records the size of the proof and updates the usage.
+	///
+	/// # Errors
+	///
+	/// Returns `ResourceError::LimitExceeded` if the proof size exceeds the limit.
+	pub fn record_proof_size(&mut self, size: u64) -> Result<(), ResourceError> {
+		self.resource.record_cost(size)
+	}
+
+	/// Refunds the given amount of proof size.
+	pub fn refund(&mut self, amount: u64) {
+		self.resource.refund(amount)
+	}
+
+	/// Returns the proof size usage.
+	pub fn usage(&self) -> u64 {
+		self.resource.usage()
+	}
+
+	/// Returns the proof size limit.
+	pub fn limit(&self) -> u64 {
+		self.resource.limit
```

Comment:
Why have an accessor for `usage()` but not for `limit`?

Reply:
Please check new version